In [1]:
import pandas as pd
df=pd.read_csv("/dataset/customer_shopping_behavior.csv",sep=";")

In [2]:
df.head()

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,Review Rating,Subscription Status,Shipping Type,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,3.1,Yes,Express,Yes,Yes,14,Venmo,Fortnightly
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,3.1,Yes,Express,Yes,Yes,2,Cash,Fortnightly
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,3.1,Yes,Free Shipping,Yes,Yes,23,Credit Card,Weekly
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,3.5,Yes,Next Day Air,Yes,Yes,49,PayPal,Weekly
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,2.7,Yes,Free Shipping,Yes,Yes,31,PayPal,Annually


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3900 entries, 0 to 3899
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Customer ID             3900 non-null   int64  
 1   Age                     3900 non-null   int64  
 2   Gender                  3900 non-null   object 
 3   Item Purchased          3900 non-null   object 
 4   Category                3900 non-null   object 
 5   Purchase Amount (USD)   3900 non-null   int64  
 6   Location                3900 non-null   object 
 7   Size                    3900 non-null   object 
 8   Color                   3900 non-null   object 
 9   Season                  3900 non-null   object 
 10  Review Rating           3863 non-null   float64
 11  Subscription Status     3900 non-null   object 
 12  Shipping Type           3900 non-null   object 
 13  Discount Applied        3900 non-null   object 
 14  Promo Code Used         3900 non-null   

In [4]:
df.describe()

,Customer ID,Age,Purchase Amount (USD),Review Rating,Previous Purchases
count,3900.000000,3900.000000,3900.000000,3863.000000,3900.000000
mean,1950.500000,44.068462,59.764359,3.750065,25.351538
std,1125.977353,15.207589,23.685392,0.716983,14.447125
min,1.000000,18.000000,20.000000,2.500000,1.000000
25%,975.750000,31.000000,39.000000,3.100000,13.000000
50%,1950.500000,44.000000,60.000000,3.800000,25.000000
75%,2925.250000,57.000000,81.000000,4.400000,38.000000
max,3900.000000,70.000000,100.000000,5.000000,50.000000


In [5]:
df.isnull().sum()

Customer ID                0
Age                        0
Gender                     0
Item Purchased             0
Category                   0
Purchase Amount (USD)      0
Location                   0
Size                       0
Color                      0
Season                     0
Review Rating             37
Subscription Status        0
Shipping Type              0
Discount Applied           0
Promo Code Used            0
Previous Purchases         0
Payment Method             0
Frequency of Purchases     0
dtype: int64

In [6]:
df['Review Rating']=df.groupby('Category')['Review Rating'].transform(lambda x:x.fillna(x.median()))

In [7]:
df.isnull().sum()

Customer ID               0
Age                       0
Gender                    0
Item Purchased            0
Category                  0
Purchase Amount (USD)     0
Location                  0
Size                      0
Color                     0
Season                    0
Review Rating             0
Subscription Status       0
Shipping Type             0
Discount Applied          0
Promo Code Used           0
Previous Purchases        0
Payment Method            0
Frequency of Purchases    0
dtype: int64

In [8]:
df.columns=df.columns.str.lower()
df.columns=df.columns.str.replace(' ','_')
df=df.rename(columns={'purchase_amount_(usd)' : 'purchase_amount'})

In [9]:
df.columns

Index(['customer_id', 'age', 'gender', 'item_purchased', 'category',
       'purchase_amount', 'location', 'size', 'color', 'season',
       'review_rating', 'subscription_status', 'shipping_type',
       'discount_applied', 'promo_code_used', 'previous_purchases',
       'payment_method', 'frequency_of_purchases'],
      dtype='object')

In [10]:
labels=['young_adults','adults','middle_aged','senior']
df['age_group']=pd.qcut(df['age'],q=4,labels=labels)

In [11]:
df[['age','age_group']]

,age,age_group
0,55,middle_aged
1,19,young_adults
2,50,middle_aged
3,21,young_adults
4,45,middle_aged
...,...,...
3895,40,adults
3896,52,middle_aged
3897,46,middle_aged
3898,44,adults


In [12]:
frequency_mapping={
    'Fortnightly':14,
    'Weekly':7,
    'Monthly':30,
    'Quarterly':90,
    'Bi-Weekly':14,
    'Annually':365,
    'Every 3 Months':90
}
df['Purchase_frequency_days']=df['frequency_of_purchases'].map(frequency_mapping)

In [13]:
df[['Purchase_frequency_days','frequency_of_purchases']]

,Purchase_frequency_days,frequency_of_purchases
0,14,Fortnightly
1,14,Fortnightly
2,7,Weekly
3,7,Weekly
4,365,Annually
...,...,...
3895,7,Weekly
3896,14,Bi-Weekly
3897,90,Quarterly
3898,7,Weekly


In [14]:
df[['discount_applied','promo_code_used']]

,discount_applied,promo_code_used
0,Yes,Yes
1,Yes,Yes
2,Yes,Yes
3,Yes,Yes
4,Yes,Yes
...,...,...
3895,No,No
3896,No,No
3897,No,No
3898,No,No


In [15]:
(df['discount_applied']==df['promo_code_used']).all()

np.True_

In [16]:
df=df.drop('promo_code_used',axis=1)

In [17]:
df.columns

Index(['customer_id', 'age', 'gender', 'item_purchased', 'category',
       'purchase_amount', 'location', 'size', 'color', 'season',
       'review_rating', 'subscription_status', 'shipping_type',
       'discount_applied', 'previous_purchases', 'payment_method',
       'frequency_of_purchases', 'age_group', 'Purchase_frequency_days'],
      dtype='object')

In [18]:
pip install psycopg2-binary sqlalchemy

Note: you may need to restart the kernel to use updated packages.


## Hypothesis Testing

We test three business relevant hypotheses using statistical tests to validate or challenge assumptions in the data.

---

### Hypothesis 1 Do discount users spend more than non-discount users?
**H₀:** Mean purchase amount is the same regardless of discount usage  
**H₁:** Customers who used a discount spend more on average  
**Test:** Independent samples t-test (two-tailed, α = 0.05)

In [21]:
from scipy import stats

discount_yes = df[df['discount_applied'] == 'Yes']['purchase_amount']
discount_no  = df[df['discount_applied'] == 'No']['purchase_amount']

t_stat, p_val = stats.ttest_ind(discount_yes, discount_no, equal_var=False)

print(f"discount group Mean spend: ${discount_yes.mean():.2f}, n={len(discount_yes)}")
print(f"no discount group Mean spend: ${discount_no.mean():.2f}, n={len(discount_no)}")
print(f"\nwelch's t-statistic : {t_stat:.4f}")
print(f"p-value             : {p_val:.4f}")
print()
if p_val < 0.05:
    print("Result: REJECT H₀ -There is a statistically significant difference in spend between discount and non-discount customers.")
else:
    print("Result: FAIL TO REJECT H₀ -No statistically significant difference in spend between the two groups.")

discount group Mean spend: $59.28, n=1677
no discount group Mean spend: $60.13, n=2223

welch's t-statistic : -1.1122
p-value             : 0.2661

Result: FAIL TO REJECT H₀ -No statistically significant difference in spend between the two groups.


Discount customers and non-discount customers spend almost the same amount on average ( 59.28 vs $60.13). Discounts are not driving higher transaction value in this dataset

---

### Hypothesis 2 Do younger customers (young_adults) buy more frequently than seniors?
**H₀:** Median purchase frequency (days between purchases) is the same for young adults and seniors  
**H₁:** Young adults purchase more frequently (fewer days between purchases)  
**Test:** Mann-Whitney U (non-parametric, since frequency_days is ordinal-mapped)

In [22]:
young = df[df['age_group'] == 'young_adults']['Purchase_frequency_days']
senior = df[df['age_group'] == 'senior']['Purchase_frequency_days']

u_stat, p_val2 = stats.mannwhitneyu(young, senior, alternative='less')

print(f"young adults Median frequency: {young.median():.0f} days, n={len(young)}")
print(f"seniors Median frequency: {senior.median():.0f} days, n={len(senior)}")
print(f"\nmann-whitney U statistic : {u_stat:.0f}")
print(f"p-value                  : {p_val2:.4f}")
print()
if p_val2 < 0.05:
    print("Result: REJECT H₀ -Young adults purchase significantly more frequently than seniors.")
else:
    print("Result: FAIL TO REJECT H₀ - No significant difference in purchase frequency between age groups.")

young adults Median frequency: 30 days, n=1028
seniors Median frequency: 30 days, n=944

mann-whitney U statistic : 469190
p-value                  : 0.0954

Result: FAIL TO REJECT H₀ - No significant difference in purchase frequency between age groups.


Gender does not significantly predict product category preference in this dataset. Marketing segmentation by gender may not be effective for this retailer.

---

### Hypothesis 3 Is there an association between gender and product category preference?
**H₀:** Gender and product category are independent  
**H₁:** There is a statistically significant association between gender and category  
**Test:** Chi-square test of independence (α = 0.05)

In [23]:
contingency = pd.crosstab(df['gender'], df['category'])
chi2, p_val3, dof, expected = stats.chi2_contingency(contingency)

print("Observed counts (Gender × Category):")
print(contingency)
print(f"\nChi-square statistic : {chi2:.4f}")
print(f"Degrees of freedom   : {dof}")
print(f"p-value              : {p_val3:.4f}")
print()
if p_val3 < 0.05:
    print("Result: REJECT H₀ — Gender and product category are NOT independent; purchasing patterns differ by gender.")
else:
    print("Result: FAIL TO REJECT H₀ — No significant association between gender and product category.")

Observed counts (Gender × Category):
category  Accessories  Clothing  Footwear  Outerwear
gender                                              
Female            392       556       199        101
Male              848      1181       400        223

Chi-square statistic : 0.5984
Degrees of freedom   : 3
p-value              : 0.8968

Result: FAIL TO REJECT H₀ — No significant association between gender and product category.


All three hypotheses were not supported by the data. This suggests demographic variables alone are insufficient for segmentation a clustering approach on behavioral variables (spend, frequency, category) would likely yield more actionable segments."

---